src/projections/igh_o.cpp

```diff
@@ -0,0 +1,235 @@
+                       (lp.phi  >=   d60-EPSLN && lp.phi <=   d90+EPSLN)); break;
+        case  2:  ok = (lp.lam  >= -d110-EPSLN && lp.lam <=   d50+EPSLN) ||
+                       ((lp.lam >=   d50-EPSLN && lp.lam <=   d60+EPSLN) &&
+                       (lp.phi  >=   d60-EPSLN && lp.phi <=   d90+EPSLN)) ||
+                       ((lp.lam <= -d120-EPSLN && lp.lam >= -d110+EPSLN) &&
+                       (lp.phi  >=   d60-EPSLN && lp.phi <=   d90+EPSLN)) ; break;
+        case  3:  ok = (lp.lam  >=   d50-EPSLN && lp.lam <=  d180+EPSLN) ||
+                       ((lp.lam >=   d10-EPSLN && lp.lam <=   d60+EPSLN) &&
+                       (lp.phi  >=   d60-EPSLN && lp.phi <=   d90+EPSLN)); break;
+        case  4:  ok = (lp.lam  >= -d180-EPSLN && lp.lam <= -d110+EPSLN); break;
+        case  5:  ok = (lp.lam  >= -d110-EPSLN && lp.lam <=   d50+EPSLN); break;
+        case  6:  ok = (lp.lam  >=   d50-EPSLN && lp.lam <=  d180+EPSLN); break;
+        case  7:  ok = (lp.lam  >= -d180-EPSLN && lp.lam <=  -d75+EPSLN); break;
+        case  8:  ok = (lp.lam  >=  -d75-EPSLN && lp.lam <=   d90+EPSLN); break;
+        case  9:  ok = (lp.lam  >=   d90-EPSLN && lp.lam <=  d180+EPSLN); break;
+        case  10: ok = (lp.lam  >= -d180-EPSLN && lp.lam <=  -d75+EPSLN); break;
+        case  11: ok = (lp.lam  >=  -d75-EPSLN && lp.lam <=   d90+EPSLN); break;
+        case  12: ok = (lp.lam  >=   d90-EPSLN && lp.lam <=  d180+EPSLN); break;
+
+        }
+      z = (!ok? 0: z); /* projectable? */
+    }
+
+    if (!z) lp.lam = HUGE_VAL;
+    if (!z) lp.phi = HUGE_VAL;
+
+    return lp;
+}
+
+
+static PJ *destructor (PJ *P, int errlev) {
+    int i;
+    if (nullptr==P)
+        return nullptr;
+
+    if (nullptr==P->opaque)
+        return pj_default_destructor (P, errlev);
+
+    for (i = 0; i < 12; ++i) {
+        if (static_cast<struct pj_opaque*>(P->opaque)->pj[i])
+            static_cast<struct pj_opaque*>(P->opaque)->pj[i]->destructor(static_cast<struct pj_opaque*>(P->opaque)->pj[i], errlev);
+    }
+
+    return pj_default_destructor(P, errlev);
+}
+
+
+
+/*
+  Zones:
+
+    -180      -110               50           180
+      +---------+----------------+-------------+    Zones 1,2,3,10,11 & 12:
+      |1        |2               |3            |    Mollweide projection
+      |         |                |             |
+      +---------+----------------+-------------+    Zones 4,5,6,7,8 & 9:
+      |4        |5               |6            |    Sinusoidal projection
+      |         |                |             |
+    0 +---------+--+-------------+--+----------+
+      |7           |8               |9         |
+      |            |                |          |
+      +------------+----------------+----------+
+      |10          |11              |12        |
+      |            |                |          |
+      +------------+----------------+----------+
+    -180          -75               90        180
+*/
+
+#define SETUP(n, proj, x_0, y_0, lon_0) \
```

Comment:
This looks like an ancient PROJ.4 macro... Surely this can be redefine as an actual function that reads better

Reply:
I am not really a C++ programmer - could we leave this as an issue to for someone more experienced than me to revisit both igh* projections at a later date?

Reply:
We can, but the likelihood of fixing this is close to zero I reckon. Honestly, this is not code I am happy to maintain. Who do you suggest revisit this?

Reply:
This is a bit beyond my scope in C++.

Reply:
Something along those lines:
```
static void setup_zone(struct pj_opaque *Q, int n,  PJ*(*proj_ptr)(PJ*), double x_0, double y_0, double lon_0
{
    if (!(Q->pj[n-1] = proj_ptr(nullptr))) return false;
    if (!(Q->pj[n-1] = proj_ptr(Q->pj[n-1]))) return false;
    Q->pj[n-1]->ctx = P->ctx;
    Q->pj[n-1]->x0 = x_0;
    Q->pj[n-1]->y0 = y_0;
    Q->pj[n-1]->lam0 = lon_0;
    return true;
}
```

and then replace call sites like:
```
if (!setup_zone(Q, 4, pj_sinu, -d150, 0, -d150) ||
    !setup_zone(Q, 5, pj_sinu, -d40,  0,  -d40) ||
   ... add others ... )
{
   return destructor(P, ENOMEM);
}
```

Reply:
Sorry - still having trouble following.  Theres this block in the middle of the SETUP() calls that adjusts the Northern Hemisphere edge zones:

```
244     /* y0 ? */
245     xy1 = Q->pj[0]->fwd(lp, Q->pj[0]); /* zone 1 */
246     xy3 = Q->pj[4]->fwd(lp, Q->pj[4]); /* zone 3 */
247     /* y0 + xy1.y = xy3.y for lt = 40d44'11.8" */
248     Q->dy0 = xy3.y - xy1.y;
249
250     Q->pj[0]->y0 = Q->dy0;
```

Reply:
> Sorry - still having trouble following. Theres this block in the middle of the SETUP()

what's the issue ? You can let this block unchanged. I just suggested to use a function instead of the SETUP macro

Reply:
Sorry - I think I see it now ...